### ISSUE TITLE
RFC: decide! tactic variant

### ISSUE THREAD

**Issue by nomeata** [leanprover/lean4#5629]
### Proposal

In the `equational_theories` project there are many script-generated theorems with essentially `by decide` as the proof. In theses cases, we really know that the `decide` tactic will succeed (if not there is a bug somewhere), and it’s silly for `decide` to first run the decision procedure at elaboration time, and then again in the kernel.

Thus I propose to add a `decide!` variant of the tactic that causes this check to be skipped:

https://github.com/leanprover/lean4/blob/01d414ac36dc28f3e424dabd36d818873fea655c/src/Lean/Elab/Tactic/ElabTerm.lean#L404-L405

This halfs the processing time of such files.

(The equational_theories project is currently using a tactic that has this effect.)

### Community Feedback

(none yet)

### Impact

Add :+1: to [issues you consider important](https://github.com/leanprover/lean4/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc). If others benefit from the changes in this proposal being added, please ask them to add :+1: to it.

**Comment by hrmacbeth** [comment]
There's an inconsistency here with the meaning of `!` in the mathlib3 `decide` tactic (which was there called `dec_trivial`).  In [that tactic](https://github.com/leanprover-community/mathlib3/blob/master/src/tactic/dec_trivial.lean), `!` meant "revert everything first and then run decide".  (Incidentally, I still miss that feature, it would be nice to have it in `decide`!)

This need not block using `!` differently in Lean 4, just wanted to ensure it was a conscious choice.

**Comment by nomeata** [comment]
Thanks, I didn’t know that! Maybe it’s generally a bad habit to just reach for `!` whenever one creates a variant of a tactic. Maybe a word like `unchecked_decide` is better.

**Comment by kmill** [comment]
With #5665 we now have a version of `decide!` that satisfies the motivation for this RFC — should we close it?

There are some ways that `decide!` really is stronger than `decide`, namely that it uses kernel reduction so can unfold everything in sight. Some possibilities to address `!` and the old Lean 3 functionality:
* add a `decide (revert := true)` option to both `decide` and `decide!`
* or, rename `decide!` to `kernel_decide` and use a `!` suffix to mean "revert the context" for both `decide!` and `kernel_decide!`.

**Comment by nomeata** [comment]
Yes, please close this (and feel free to do it yourself, so that credit goes where credit is due). I thought I edited that PRs description so that it would do it, but maybe didn't save.

I don't know much about the lean3 feature. Are people asking for it? Could it be the default behavior to always revert (maybe with a good explanation that this happened if no instance is found)?

**Comment by kmill** [comment]
@nomeata There are some cases where `decide` could succeed if it reverted variables the goal depends on, but I've gone back and forth whether this would be a good default. Maybe better would be a compositional approach, with a `revert_deps` tactic to revert just dependencies?

**Comment by nomeata** [comment]
Thats also plausible. Are there situations when reverting is _not_ the right thing to do for `decide`?
